### PR TITLE
Add postgres-specific id generator

### PR DESCRIPTION
**Pull request by 4word**
Requiring an external-to-the-db id generator is not an option for some use cases. This new field and related migration utilities allow for shard-safe (up to 8191 shards) inserts with guaranteed unique IDs (up to 1024 inserts per millisecond on an individual shard). The ID will return like a standard Django auto field ID. In the future, a given ID could be deconstructed to get the time it was inserted and the shard id (useful for the router to potentially know which shard to grab something from without needing help from the using() statement).

### CODE REVIEW

**Comment by 4word** [comment]
Note that much of this code was ported over from a Django 1.4 and python 2.7 deployment and I am not exactly expecting unit tests to pass the first time (although I updated as much as I could think of and find).

**Comment by 4word** [comment]
This is ready for more scrutinizing review. Docs need to be updated but functionality is available and tested for both database and django level usage.

File: django_sharding_library/fields.py [hunk 1]
```diff
@@ -156,3 +159,30 @@ class ShardForeignKeyStorageField(ShardForeignKeyStorageFieldMixin, ForeignKey):
...
+            return super(PostgresShardGeneratedIDField, self).db_type(connection)
+
+    @staticmethod
+    def migration_receiver(*args, **kwargs):
+        sequence_name = "global_id_sequence"
+        db_alias = kwargs.get('using')
+        if not db_alias:
+            raise EnvironmentError("A pre-migration receiver did not receive a database alias. "
+                                   "Perhaps your app is not registered correctly?")
+        if settings.DATABASES[db_alias]['ENGINE'] == Backends.POSTGRES:
```
Comment (JBKahn) [inline django_sharding_library/fields.py]: since this only supports postgres, somewhere (maybe here?) we should raise an exception if you try to use it and you're not using postgres.

File: django_sharding_library/utils.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+def create_postgres_shard_id_function(sequence_name, db_alias, shard_id):
+    cursor = connections[db_alias].cursor()
+    cursor.execute(postgres_shard_id_function_sql % {'shard_epoch': settings.SHARD_EPOCH,
+                                                     'shard_id': shard_id,
+                                                     'sequence_name': sequence_name})
+    cursor.close()
+
+
+def register_migration_signal_for_model_receiver(model, function, dispatch_uid=None):
+    signals.pre_migrate.connect(function, sender=model, dispatch_uid=dispatch_uid)
```
Comment (JBKahn) [inline django_sharding_library/utils.py]: why pre_migrate? Is there a check that it was successful or a chance there will be a silent failure such that one node doesn't have it?

File: django_sharding_library/fields.py (hunk 1 above)
Comment (4word) [inline django_sharding_library/fields.py]: Hmm. The problem is that the library allows arbitrary backends within the same shard group. Should I just loop through the entire shard group and make sure all the backends are postgres?

File: django_sharding_library/utils.py (hunk 2 above)
Comment (4word) [inline django_sharding_library/utils.py]: It will get a DatabaseError if it fails. Its in pre-migrate because its the only common area to add a migration step without editing the migration file itself after its produced, which I don't think is a good feature for a library since it adds complexity to using the library that can be taken care of automatically in the pre migration.

Also, it has to be in the pre migrate because if one of the postgres fields gets created and the stored procedure doesnt exist yet, it will raise a programmingerror. Doing it in a post migration or anywhere else doesn't work.

File: django_sharding_library/fields.py (hunk 1 above)
Comment (JBKahn) [inline django_sharding_library/fields.py]: Yup, you can do it either for all sharded databases or make it a shard
group level setting?

File: django_sharding_library/utils.py (hunk 2 above)
Comment (JBKahn) [inline django_sharding_library/utils.py]: Hmm definitely don't think we need to force users to modify sharded tables.
If we can, I have to look into how database migration dependencies work but
could use a migration dependency on a migration shipped with the app?

Comment (4word) [inline django_sharding_library/utils.py]: As far as I can find in the documentation, there is no way to force a dependency because of a field when a user run makemigrations. I also asked in the django irc and the response from the admin there at the time was that he didn't think it was possible.

This is exactly why the pre-migration signal was added in though, its for pre-migration processing. I am pretty sure this is the "django" way of doing it short of writing a migration, putting it in the app, and figuring out how to force it as a dependency on another app based on arbitrary, non-built-in logic (for example, foreign key fields have special logic already built into the migration builder). I am not sure I can think of a better use case of the pre migration signal than this

**Comment by 4word** [comment]
Updated the decorator to raise an error when the postgres field is used on a non-postgres database.

File: django_sharding_library/fields.py (hunk 1 above)
Comment (4word) [inline django_sharding_library/fields.py]: Updated to raise an exception

**Comment by 4word** [comment]
Fixed a bug where when there are multiple models in the same app, and some of them do not use postgres, the migration was trying to run the postgres-specific functions on the non-postgres models. Its fixed to check for postgres correctly now.

File: django_sharding_library/utils.py (hunk 2 above)
Comment (4word) [inline django_sharding_library/utils.py]: Let me know if you find something to handle a dependency, I actually already have a migration written to do this that I could include if we figure it out (in my current project, I was adding the migration manually to the automated migration files before I wrote this)

**Comment by JBKahn** [comment]
I'm going to spend one more day stewing on it, the dependancy code didn't seem to offer a lot of ways to reasonably hook into it.

File: django_sharding_library/decorators.py [hunk 3]
```diff
@@ -24,12 +30,30 @@ def configure(cls):
                 )
             setattr(cls, 'django_sharding__database', database)
 
+        postgres_shard_id_fields = list(filter(lambda field: issubclass(type(field), PostgresShardGeneratedIDField), cls._meta.fields))
+        if postgres_shard_id_fields:
+            # Make sure we are only using postgres
+            if database:
```
Comment (JBKahn) [inline django_sharding_library/decorators.py]: I'd just refactor that check into something like:
````python
if postgres_shard_id_fields:
    database_dicts = [settings.DATABASES[database]] if database else [db_settings for db, db_settings in viewitems(settings.DATABASES) if db_settings["SHARD_GROUP"] == shard_group]
    if any([database_dict['ENGINE'] not in Backends.POSTGRES for database_dict in database_dicts]):
        raise ShardedModelInitializationException('You cannot use a PostgresShardGeneratedIDField on a non-Postgres database.')

File: django_sharding_library/decorators.py [hunk 4]
```diff
@@ -24,12 +30,30 @@ def configure(cls):
...
+                    raise ShardedModelInitializationException('You cannot use a PostgresShardGeneratedIDField on a '
+                                                              'non-Postgres database.')
+
+            if shard_group:
+                for db_key in settings.DATABASES.keys():
+                    if settings.DATABASES[db_key]['ENGINE'] not in Backends.POSTGRES:
+                        raise ShardedModelInitializationException('You cannot use a PostgresShardGeneratedIDField on a '
+                                                                  'non-Postgres database.')
+
+            register_migration_signal_for_model_receiver(apps.get_app_config(cls._meta.app_label),
```
Comment (JBKahn) [inline django_sharding_library/decorators.py]: I spent more time reading that code for dependencies and it isn't worth trying to hack into it, given they don't expose any good hooks for us to use.

File: django_sharding_library/decorators.py [hunk 5]
```diff
@@ -24,12 +30,30 @@ def configure(cls):
...
+                                                         PostgresShardGeneratedIDField.migration_receiver,
+                                                         dispatch_uid=PRE_MIGRATION_DISPATCH_UID % cls._meta.app_label)
+
         if shard_group:
             sharded_fields = list(filter(lambda field: issubclass(type(field), ShardedIDFieldMixin), cls._meta.fields))
-            if not sharded_fields:
+            if not sharded_fields and not postgres_shard_id_fields:
```
Comment (JBKahn) [inline django_sharding_library/decorators.py]: update those two exception messages to include that using the `PostgresShardGeneratedIDField` is also valid.

File: django_sharding_library/fields.py [hunk 6]
```diff
@@ -156,3 +159,30 @@ class ShardForeignKeyStorageField(ShardForeignKeyStorageFieldMixin, ForeignKey):
...
+            raise ValueError("PostgresShardGeneratedIDField requires a SHARD_EPOCH to be defined in your settings file.")
+
+        if connection.vendor == PostgresDatabaseWrapper.vendor:
+            return "bigint DEFAULT next_sharded_id()"
+        else:
+            return super(PostgresShardGeneratedIDField, self).db_type(connection)
+
+    @staticmethod
+    def migration_receiver(*args, **kwargs):
+        sequence_name = "global_id_sequence"
```
Comment (JBKahn) [inline django_sharding_library/fields.py]: Ohh it's one sequence per DB,, as opposed to per model?

File: django_sharding_library/settings_helpers.py [hunk 7]
```diff
@@ -89,4 +90,13 @@ def database_configs(databases_dict):
                 )
                 if db_config:
                     configuration[replica['name']] = db_config
+
+            # We assume the numeric shard ID is constant based on the entries in the configuration helper (we assume
+            # they wont change order, and that new shards will be appended and not inserted randomly)
+            # todo: update docs to make sure this assumption is understood
```
Comment (JBKahn) [inline django_sharding_library/settings_helpers.py]: very important if we make that assumption.

File: django_sharding_library/utils.py (hunk 2 above)
Comment (JBKahn) [inline django_sharding_library/utils.py]: Like I said, I have no ideas, make a special note in your docs about this. Having to run this. Ideally I'd also like to add a management command to run this, as well as one to check that ti was run. Since the django way feels a bit hacky to me.

File: django_sharding_library/fields.py (hunk 6 above)
Comment (4word) [inline django_sharding_library/fields.py]: Correct, its per DB. There is no reason to have a per model, no DB (that I am aware of) can insert more than 1024 rows per millisecond (which is what this id generator is designed to be able to do)!

File: django_sharding_library/decorators.py (hunk 5 above)
Comment (4word) [inline django_sharding_library/decorators.py]: Will do that today

File: django_sharding_library/decorators.py (hunk 4 above)
Comment (4word) [inline django_sharding_library/decorators.py]: Yeah its unfortunate that django has opted to keep migrations behind a closed curtain it seems like. Will look at updating the docs today.

File: django_sharding_library/decorators.py (hunk 3 above)
Comment (4word) [inline django_sharding_library/decorators.py]: Good call, that looks cleaner

File: django_sharding_library/settings_helpers.py (hunk 7 above)
Comment (4word) [inline django_sharding_library/settings_helpers.py]: I will make sure to have that in the doc

**Comment by 4word** [comment]
Will be trying to wrap this up later today, I tried to get to it earlier but it was really busy this week.

File: django_sharding_library/utils.py (hunk 2 above)
Comment (4word) [inline django_sharding_library/utils.py]: I updated the docs to indicate this is happening. I can add a migrate command in the future: this is pretty failure-safe. If for any reason the required function or sequence do not exist, postgres itself will error out and send the error message to Django. But I agree, a manage command is in order, at least for checking that the function exists and is working properly. Will do that in a future pull request.

File: django_sharding_library/settings_helpers.py (hunk 7 above)
Comment (4word) [inline django_sharding_library/settings_helpers.py]: Doc updated with this assumption

Comment (JBKahn) [inline django_sharding_library/settings_helpers.py]: Awesome, a bit out of commission for a few days but later this week I'll
get the Travis pull request merged as well as take another look at your code

File: django_sharding_library/fields.py (hunk 6 above)
Comment (JBKahn) [inline django_sharding_library/fields.py]: when we were deciding on that strategy, we just weren't sure about id growth and sharing them, although now that I consider the actual size of a bigint, I have no idea why the decision was made not to do that.

File: django_sharding_library/utils.py (hunk 2 above)
Comment (JBKahn) [inline django_sharding_library/utils.py]: add the management command in this one, at least a rudimentary one.
i.e.
```python
ipdb> cursor.execute("SELECT count(*) FROM pg_class c WHERE c.relkind = '%s' and c.relname = '%s';" % ('S', sequence_name))
ipdb> cursor.fetchone()
(1L,)
```
and just take a sequence name (or whatever requirements to generate the ones you use) and the database to check. As well as one to create the sequence and to reset it, using the code you already wrote.

Comment (JBKahn) [inline django_sharding_library/utils.py]: well I supose it can be in a different PR but I don't want to cut a release till it's in.

**Comment by JBKahn** [comment]
The code looks good, two questions:
1) Is there an upper limit to the number of shards given they are assigned ids, although I assume it's very large in either event?
2) If the shard epoch can never be changed anyway, is there a reason we don't set a default that we let you override? I assume that unless you're moving from basically the same system to this one, you'd never care about the value?

**Comment by 4word** [comment]
There is an upper limit, it's stored in 10 bits which limits it to 8192 (minus 1, since the code assumes you start your shard numbering at 1 instead of 0) total shards. I will put that in the documentation.

I can put a default in there, the main thing is that anyone writing code in the future needs to never change the default value because it is backwards incompatible, anyone who starts using the library with the default will never be able to change it without migrating all their data.

Because of the gravity of the potential amount of work involved with changing the epoch, I would rather just raise an exception and make the person make a decision about it. They can use one of the other ID generators in the library if they don't want to think about it.

**Comment by JBKahn** [comment]
that's reasonable. I'll merge this one in, but I'll hold off on the release till the management command is in since I'd rather group those into a release.

**Comment by JBKahn** [comment]
Well, after you modify the docs :)

**Comment by 4word** [comment]
That sounds good. I'll update the docs later today.

**Comment by 4word** [comment]
I rebased my commits to squash down the sheer volume of them that there were. I updated the doc to mention the shard limit.

This is ready to merge in, I am planning on doing the management commands in a new PR. I would rather wait for you to merge this in until after you release the manager and router changes on pypi, if you are planning on doing that soon, since its useful for most people. If not, then just merge it whenever and I will plan on working on the management commands later this week.

File: docs/components/IDGeneration.md [hunk 8]
```diff
@@ -22,6 +22,7 @@ The two included in the package are:
```
Comment (bronsen) [inline docs/components/IDGeneration.md]: Shouldn't it say
> The _three_ included in the package are


Also, I wonder if one could clarify what is enumerated here, since all these are mentioned above: strategies, additional arguments, or requirements. (<-- this is me being very pedantic)

Comment (4word) [inline docs/components/IDGeneration.md]: You are right, it should say three. I will update.

What do you mean clarify what is enumerated? There are no additional requirements other than what is listed, which basically includes Postgres and defining a shard epoch time. All of the exact requirements are in the Using section for this feature, which is in the ShardingAModel.md file. If thats not what you mean, let me know!

**Comment by JBKahn** [comment]
I'm going to try looking at this tonight. I'll probably take a stab at writing the management commands I'd like to include and then we can see about releasing this ASAP since I've been really lazy about this.

**Comment by JBKahn** [comment]
@4word if I write some commands, and this is reviewed, are any pieces missing or is this good to go (both from the point of view of the pull request and for a new release) from your point o f view?

**Comment by 4word** [comment]
All the pieces are there, should be good to be merged in.

File: django_sharding_library/fields.py [hunk 9]
```diff
@@ -156,3 +166,64 @@ class ShardForeignKeyStorageField(ShardForeignKeyStorageFieldMixin, ForeignKey):
...
+class PostgresShardForeignKey(ForeignKey):
+    def db_type(self, connection):
+        # The database column type of a ForeignKey is the column type
+        # of the field to which it points. An exception is if the ForeignKey
+        # points to an AutoField/PositiveIntegerField/PositiveSmallIntegerField,
+        # in which case the column type is simply that of an IntegerField.
+        # If the database needs similar types for key fields however, the only
+        # thing we can do is making AutoField an IntegerField.
+        rel_field = self.target_field
+        if rel_field.get_internal_type() is "BigIntegerField":
```
Comment (JBKahn) [inline django_sharding_library/fields.py]: Did you mean to put `is` and not `==` for the string comparison?

File: django_sharding_library/fields.py [hunk 10]
```diff
@@ -156,3 +166,64 @@ class ShardForeignKeyStorageField(ShardForeignKeyStorageFieldMixin, ForeignKey):
...
+class PostgresShardOneToOne(OneToOneField):
+    def db_type(self, connection):
+        # The database column type of a ForeignKey is the column type
+        # of the field to which it points. An exception is if the ForeignKey
+        # points to an AutoField/PositiveIntegerField/PositiveSmallIntegerField,
+        # in which case the column type is simply that of an IntegerField.
+        # If the database needs similar types for key fields however, the only
+        # thing we can do is making AutoField an IntegerField.
+        rel_field = self.target_field
+        if rel_field.get_internal_type() is "BigIntegerField":
```
Comment (JBKahn) [inline django_sharding_library/fields.py]: ditto

**Comment by JBKahn** [comment]
@4word those are my last two questions, then I should be good to merge this in and write the commands.

**Comment by JBKahn** [comment]
I'm going to merge and make the change and let me know if I should revert it. That way I can continue working on it before you get back to me.
